miniapps/navier/navier_particles.hpp

```diff
@@ -0,0 +1,313 @@
+                                        const Vector &s1_end,
+                                        const Vector &s2_start,
+                                        const Vector &s2_end,
+                                        Vector &x_int, real_t *t1_ptr=nullptr,
+                                        real_t *t2_ptr=nullptr);
+
+   /// Apply 2D reflection BCs to update particle positions and velocities
+   void Apply2DReflectionBC(const ReflectionBC_2D &bc);
+   /// Apply 2D recirculation BCs
+   void Apply2DRecirculationBC(const RecirculationBC_2D &bc);
+   /// Apply all BCs in \ref bcs
+   void ApplyBCs();
+
+   /** @brief Move any particles that have left the domain to the
+    *  inactive ParticleSet.
+    *
+    *  This method uses the FindPointsGSLIB object internal to the class to
+    *  detect if particles are within the domain or not.
+    *
+    *  @param[in] findpts     If true, call FindPointsGSLIB::FindPoints prior
+    *  to deactivation (if particle coordinates out of sync with
+    *  FindPointsGSLIB)
+    */
+   void DeactivateLostParticles(bool findpts);
+
+   // Temporary vectors for particle computation
+   mutable Vector up, vp, xpn, xp;
+   mutable Vector r, C;
+public:
+
+   /// Initialize NavierParticles with \p num_particles using fluid mesh \p m .
+   NavierParticles(MPI_Comm comm, int num_particles, Mesh &m);
+
+   /// Set initial timestep in time history array
+   void Setup(const real_t &dt) { dthist[0] = dt; }
+
+   /** @brief Step the particles in time.
+    *
+    *  @param[in] dt
+    *  @param[in] u_gf     Fluid velocity on fluid mesh.
+    *  @param[in] w_gf     Fluid vorticity on fluid mesh.
+    */
+   void Step(const real_t &dt, const ParGridFunction &u_gf,
+             const ParGridFunction &w_gf);
+
+   /** @brief Interpolate fluid velocity and vorticity onto current particles'
+    *  location.
+    *
+    *  @param[in] u_gf     Fluid velocity on fluid mesh.
+    *  @param[in] w_gf     Fluid vorticity on fluid mesh.
+    */
+   void InterpolateUW(const ParGridFunction &u_gf, const ParGridFunction &w_gf);
+
+   /// Get reference to the active ParticleSet.
+   ParticleSet& GetParticles() { return fluid_particles; }
+
+   /// Get reference to the inactive ParticleSet.
+   ParticleSet& GetInactiveParticles() { return inactive_fluid_particles; }
+
+   /// Get reference to the kappa ParticleVector.
+   ParticleVector& Kappa()
+   {
+      return fluid_particles.Field(fp_idx.field.kappa);
+   }
+
+   /// Get reference to the Zeta ParticleVector.
+   ParticleVector& Zeta()
+   {
+      return fluid_particles.Field(fp_idx.field.zeta);
+   }
+
+   /// Get reference to the Gamma ParticleVector.
+   ParticleVector& Gamma()
+   {
+      return fluid_particles.Field(fp_idx.field.gamma);
+   }
+
+   /// Get reference to the fluid velocity-interpolated ParticleVector at
+   /// time n - \p nm .
+   ParticleVector& U(int nm=0)
+   {
+      MFEM_ASSERT(nm < 4, "nm must be <= 3");
```

Comment:
Make a static constexpr constant of the history length, instead of hardcoding it everywhere 😬 .